Several performance optimizations

paulrutter opened the pull request:
We have been using AlaSQL for a couple of years, and had some patches applied to it locally.
These patches were created after Node.js CPU profiling AlaSQL queries. This PR contains these patches. 

Performance improvements for AlaSQL
- Add AST cache
- Add regex pattern cache when using LIKE (building RegEx is expensive)
- Optimize when having lots of values in a (NOT) IN statement, using JS Set
- Remove RegEx checking for DATE function (what is it used for? It slows DATE functions down)
- When using NOW, just return the date (not sure why a date is translated to a string?)

Please let me know if these changes make sense and/or require changes.
Especially the last two changes might be there for a reason we don't know about.

I also skipped some test cases around string dates and adjusted one for `WHERE IN`, using plain string literals.
The PR is lacking new test cases, but if you're interested in these changes, we can work on these.

I couldn't get the formatter to work on Windows though, so that still has to be done.

Thank you for the time you are putting into AlaSQL!

paulrutter committed:
Performance improvements for AlaSQL
- Add pattern cache when using like (building RegEx is expensive)
- Optimize when having lots of values in a (NOT) IN statement, using JS Set
- Remove RegEx checking for DATE function
- When using NOW, just return the date (not sure why a date is translated to a string and then to a date again)

paulrutter committed:
Performance improvements for AlaSQL
- Add AST cache

paulrutter committed:
Performance improvements for AlaSQL
- Adjust or skip unit tests

paulrutter committed:
Performance improvements for AlaSQL
- Remove cache property

mathiasrw commented:
I love it!

Let me have a look at the code and merge as soon as possible

paulrutter commented:
I was under the impression that the failing unit tests where due to running on Windows (newlines), but apparantly these also fail in CI. Are these new failures? Then we need to look into these.

paulrutter committed:
Performance improvements for AlaSQL
- Fix LIKE unit tests by not using toUpperCase() anymore; use same code as before, but now with a patternCache

paulrutter commented:
@mathiasrw I fixed the unit tests for the LIKE, all unit tests succeed now locally:

```
  ............................................................................................................................................................
  ....,,,......................................,,,................,...........................................................................................
  .............................................................................................................................,,..,................,.........
  ...............................,.........................................................................................................................,,,
  ,,,,............................,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,..........................................,,,...................,,,,......,,,,........
  ..,,,,,,,,,,,,,,,...,,,...,,,,,,,,,,,,,,,,,,.........,,,,,,,,..................................,,,,,,,,,,,,........,,,..............................,,,,,,,,
  ,,.................................,,,,....,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,....,,,,,,................,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
  ,,,,,,,,,,,,,,,,,,,,,,,,,,,....,,,,,,,,.......,,,,,,..........................................,,,,,,,,,,,,,,,,,,,.........,,,,.,............................
  ...........................................................,........,,.,.,..,,,,,,,,,,......................................................................
  .......,.................................,,,,,,,,,,,,,,,,,,,,,,,,,..........................................................................................
  .....................................................,........................................................,.......,,..,,..,,..,,.........,..............
  ..,.......,..,......,,...................,..,........,,,.....,,,,,,,,,,,,,,,,...............,.,.............................................................
  ...................................................,.................

  1561 passing (4s)
  380 pending
```

paulrutter committed:
Performance improvements for AlaSQL
- Always use SET, remove old code

paulrutter committed:
Performance improvements for AlaSQL
- Made returning Date objects configurable as option, to remain backwards compatible
- Added unit test for Date objects in relevant test cases

paulrutter commented:
@mathiasrw I adjusted the skipped tests and re-enabled them again, as i made it configurable as option instead.
I extended the relevant unit tests.

Can you run the workflow again?

mathiasrw commented:
I sure can. Sorry for slow response. Lots of things going on at the moment. 

I love the performance optimizations, but we cant "just" redo tests that beforehand was working. If we will be doing changes that by some can be considered breaking (= tests are not verifying the same thing as before) than I want to pool those together and add them at the next major version bump (we got a few other tings that is also in the queue for that)

Are you OK to split this PR into so we seperate what demands a change of an existing test?

mathiasrw commented on test/test811.js:
```diff
@@ -102,7 +102,7 @@ describe('Test 811 - String / Number objects', function () {
 
 	it('5. Where In', function (done) {
-		var t1 = [{ID: new String('s1')}, {ID: new String('s2')}, {ID: new String('s3')}];
+		var t1 = [{ID: "s1"}, {ID: "s2"}, {ID: "s3"}];
```
This is a breaking change. We need to seperate that out of this PR

paulrutter replied:
I will look into it, maybe the implementation using sets can be adjusted to also allow for using new String.

paulrutter committed:
Performance improvements for AlaSQL
- Fix using new String in IN statement
- Add new test cases for using plain string literals in both IN and NOT IN

paulrutter commented:
@mathiasrw I fixed the implementation so it supports both string literals and new String.
Added new unit tests for using string literals in `IN`, and also added tests using `NOT IN`.

Can you check and run the workflow again? This way, we don't need to split of the PR.

mathiasrw commented on src/61date.js:
```diff
@@ -43,27 +43,30 @@ stdfn.OBJECT_ID = function (objid) {
 
 stdfn.DATE = function (d) {
-	if (/\d{8}/.test(d)) return new Date(+d.substr(0, 4), +d.substr(4, 2) - 1, +d.substr(6, 2));
+	if (!isNaN(d) && d.length === 8) return new Date(+d.substr(0, 4), +d.substr(4, 2) - 1, +d.substr(6, 2));
```
Good call

mathiasrw commented on src/50expression.js:
```diff
@@ -506,8 +509,17 @@
 				} else if (Array.isArray(this.right)) {
-					s = '([' + this.right.map(ref).join(',') + '].indexOf(';
-					s += 'alasql.utils.getValueOf(' + leftJS() + '))<0)';
+					// Added patch to have a better performance for when you have a lot of entries in a NOT IN statement
```
I love it

mathiasrw commented on src/50expression.js:
```diff
@@ -483,14 +483,17 @@
-						leftJS() +
-						'))>-1)';
-					//console.log(s);
+					// Added patch to have a better performance for when you have a lot of entries in an IN statement
```
Nice

mathiasrw commented on src/50expression.js:
```diff
@@ -483,14 +483,17 @@
+					const allValues = this.right.map((value) => value.value);
+					const allValuesStr = allValues.join(",");
+					if (!alasql.sets[allValuesStr]) {
+						// leverage JS Set, which is faster for lookups than arrays
```
Neat

mathiasrw commented on src/20database.js:
```diff
@@ -67,6 +67,7 @@ var Database = (alasql.Database = function (databaseid) {
 Database.prototype.resetSqlCache = function () {
 	this.sqlCache = {}; // Cache for compiled SQL statements
 	this.sqlCacheSize = 0;
+	this.astCache = {}; // Cache for AST objects
```
Makes sense

mathiasrw commented on src/17alasql.js:
```diff
@@ -246,19 +249,28 @@ alasql.dexec = function (databaseid, sql, params, cb, scope) {
-	// Create AST
-	var ast = alasql.parse(sql);
+	var ast = db.astCache[hh];
+	if (alasql.options.cache && !ast) {
```
Exactly how I hoped it would be

mathiasrw commented on src/17alasql.js:
```diff
@@ -119,12 +119,15 @@ alasql.options = {
 	loopbreak: 100000,
+
+	/** Whether GETDATE() and NOW() return dates as string. If false, then a Date object is returned */
+	dateAsString: true,
```
Nice way to default to the old way but keep the ability to run perfomant

mathiasrw commented on src/15utility.js:
```diff
@@ -1175,43 +1175,46 @@ var domEmptyChildren = (utils.domEmptyChildren = function (container) {
   @return {boolean} If value LIKE pattern ESCAPE escape
   */
-
+var patternCache = {};
```
Such a good idea. Love it.

paulrutter commented:
Thanks for merging the PR, will this be part of next release?
